File: scripts/test_pattern/test_chain_of_responsibility.py
"""Tests for the Chain of Responsibility pattern."""

import sys
import os
import unittest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'pattern'))
from chain_of_responsibility import (
    SupportRequest, Tier1Handler, Tier2Handler, Tier3Handler,
    build_default_chain
)


class TestChainOfResponsibility(unittest.TestCase):

    # ------------------------------------------------------------------
    # Basic routing
    # ------------------------------------------------------------------
    def test_level1_handled_by_tier1(self):
        result = build_default_chain().handle(SupportRequest(1, "password reset"))
        self.assertIn("Tier1", result)

    def test_level2_handled_by_tier2(self):
        result = build_default_chain().handle(SupportRequest(2, "network config"))
        self.assertIn("Tier2", result)

    def test_level3_handled_by_tier3(self):
        result = build_default_chain().handle(SupportRequest(3, "kernel panic"))
        self.assertIn("Tier3", result)

    # ------------------------------------------------------------------
    # Description preserved
    # ------------------------------------------------------------------
    def test_result_contains_description(self):
        result = build_default_chain().handle(SupportRequest(2, "unique_42"))
        self.assertIn("unique_42", result)

    # ------------------------------------------------------------------
    # Unhandled
    # ------------------------------------------------------------------
    def test_unhandled_request(self):
        result = build_default_chain().handle(SupportRequest(4, "alien invasion"))
        self.assertIn("Unhandled", result)

    # ------------------------------------------------------------------
    # Partial chain
    # ------------------------------------------------------------------
    def test_partial_chain_tier1_unhandled(self):
        t2 = Tier2Handler()
        result = t2.handle(SupportRequest(1, "basic question"))
        self.assertIn("Unhandled", result)

    def test_partial_chain_tier2_handled(self):
        t2 = Tier2Handler()
        result = t2.handle(SupportRequest(2, "medium question"))
        self.assertIn("Tier2", result)

    # ------------------------------------------------------------------
    # set_next returns handler (fluent chaining)
    # ------------------------------------------------------------------
    def test_set_next_returns_next(self):
        t1 = Tier1Handler()
        t2 = Tier2Handler()
        ret = t1.set_next(t2)
        self.assertIs(t2, ret)


if __name__ == "__main__":
    unittest.main()

File: test/pattern/test_chain_of_responsibility.cpp
#include "pattern/chain_of_responsibility.hpp"
#include <gtest/gtest.h>

using namespace pattern;

// ---------------------------------------------------------------------------
// Basic routing
// ---------------------------------------------------------------------------
TEST(ChainOfResponsibility, Level1HandledByTier1)
{
    auto chain  = buildDefaultChain();
    auto result = chain->handle({1, "password reset"});
    EXPECT_NE(std::string::npos, result.find("Tier1"));
}

TEST(ChainOfResponsibility, Level2HandledByTier2)
{
    auto chain  = buildDefaultChain();
    auto result = chain->handle({2, "network config"});
    EXPECT_NE(std::string::npos, result.find("Tier2"));
}

TEST(ChainOfResponsibility, Level3HandledByTier3)
{
    auto chain  = buildDefaultChain();
    auto result = chain->handle({3, "kernel panic"});
    EXPECT_NE(std::string::npos, result.find("Tier3"));
}

// ---------------------------------------------------------------------------
// Description is preserved in result
// ---------------------------------------------------------------------------
TEST(ChainOfResponsibility, ResultContainsDescription)
{
    auto chain  = buildDefaultChain();
    auto result = chain->handle({2, "unique_problem_42"});
    EXPECT_NE(std::string::npos, result.find("unique_problem_42"));
}

// ---------------------------------------------------------------------------
// Unhandled request (no handler covers level 4)
// ---------------------------------------------------------------------------
TEST(ChainOfResponsibility, UnhandledRequestReturnsUnhandled)
{
    auto chain  = buildDefaultChain();
    auto result = chain->handle({4, "alien invasion"});
    EXPECT_NE(std::string::npos, result.find("Unhandled"));
}

// ---------------------------------------------------------------------------
// Partial chain (only Tier2)
// ---------------------------------------------------------------------------
TEST(ChainOfResponsibility, PartialChainTier1Unhandled)
{
    auto t2     = std::unique_ptr<SupportHandler>(new Tier2Handler());
    auto result = t2->handle({1, "basic question"});
    EXPECT_NE(std::string::npos, result.find("Unhandled"));
}

TEST(ChainOfResponsibility, PartialChainTier2Handled)
{
    auto t2     = std::unique_ptr<SupportHandler>(new Tier2Handler());
    auto result = t2->handle({2, "medium question"});
    EXPECT_NE(std::string::npos, result.find("Tier2"));
}

// ---------------------------------------------------------------------------
// setNext returns the new tail for easy chaining
// ---------------------------------------------------------------------------
TEST(ChainOfResponsibility, SetNextReturnsNextPointer)
{
    auto  t1    = std::unique_ptr<SupportHandler>(new Tier1Handler());
    auto* t2raw = new Tier2Handler();
    auto* ret   = t1->setNext(std::unique_ptr<SupportHandler>(t2raw));
    EXPECT_EQ(t2raw, ret);
}

int main(int argc, char** argv)
{
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
